expression/scalar_function.go

```diff
@@ -438,13 +439,120 @@ func (sf *ScalarFunction) EvalJSON(ctx sessionctx.Context, row chunk.Row) (types
 
 // HashCode implements Expression interface.
 func (sf *ScalarFunction) HashCode(sc *stmtctx.StatementContext) []byte {
+	if sc.CanonicalHashCode.Load() {
+		simpleCanonicalizedHashCode(sf, sc)
+		return sf.hashcode
+	}
 	if len(sf.hashcode) > 0 {
 		return sf.hashcode
 	}
 	ReHashCode(sf, sc)
 	return sf.hashcode
 }
 
+// ExpressionsSemanticEqual is used to judge whether two expression tree is semantic equivalent.
+func ExpressionsSemanticEqual(ctx sessionctx.Context, expr1, expr2 Expression) bool {
+	sc := ctx.GetSessionVars().StmtCtx
+	sc.CanonicalHashCode.Store(true)
+	defer sc.CanonicalHashCode.Store(false)
+	return bytes.Equal(expr1.HashCode(sc), expr2.HashCode(sc))
```

Comment:
`(*ScalarFunction).HashCode` is not a pure function and has the side effect(i.e., if `len(sf.hashcode) > 0`, then we don't need to recalculate it). Is it ok to expose the hash code calculated under `CanonicalHashCode` to outside?

Reply:
Nice catch!
Addressed.